spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/PulsarMessageChannelBinder.java

```diff
@@ -118,6 +129,40 @@ protected MessageProducer createConsumerEndpoint(ConsumerDestination destination
 		return pulsarMessageDrivenChannelAdapter;
 	}
 
+	// VisibleForTesting
+	@Nullable
+	Schema<Object> resolveSchema(@Nullable SchemaType schemaType, @Nullable Class<?> messageType,
+			@Nullable Class<?> messageKeyType, @Nullable Class<?> messageValueType) {
+		if (schemaType == null) {
+			schemaType = SchemaType.NONE;
+		}
+		ResolvableType resolvableType = null;
+		if (schemaType.isStruct()) {
+			resolvableType = ResolvableType.forClass(Objects.requireNonNull(messageType,
+					"'message-type' required for 'schema-type' " + schemaType.name()));
+		}
+		else if (schemaType == SchemaType.KEY_VALUE) {
```

Comment:
There is a caveat w/ the `KEY_VALUE` schema resolution currently (not just here - same for template and listener)...

Complex types will map to `BYTES` unless the user adds a custom mapping for the type.  For example, given a `KevValue<String, Foo>` - it will use `Schema.STRING` and `Schema.BYTES`. 

However, if we add a custom mapping to the resolver for `Foo.class -> JSONSchema.of(Foo.class)` then it will map properly.